### ISSUE TITLE
SMB and SFTP file read broken in release build

### ISSUE THREAD

**Issue by zhanghai** [zhanghai/MaterialFiles#1154]
From Crashlytics:

```
Fatal Exception: java.lang.NoSuchFieldError: No instance field $transform of type LE5/l; in class LK6/i; or its superclasses (declaration of 'K6.i' appears in /data/app/~~h--0qtK3ADz2gHigRLL33g==/me.zhanghai.android.files-SCiTyBFzgKweE984MWA0Hg==/base.apk)
       at me.zhanghai.android.files.provider.smb.client.FileByteChannel$onReadAsync$$inlined$map$1.get(FutureExtensions.kt:1)
       at me.zhanghai.android.files.provider.common.AbstractFileByteChannel$ReadBuffer.readIntoBuffer(AbstractFileByteChannel.kt:40)
       at me.zhanghai.android.files.provider.common.AbstractFileByteChannel$ReadBuffer.read(AbstractFileByteChannel.kt:40)
       at me.zhanghai.android.files.provider.common.AbstractFileByteChannel.read(AbstractFileByteChannel.kt:26)
       at me.zhanghai.android.files.file.FileProvider$ChannelCallback.onRead(FileProvider.kt:40)
       at me.zhanghai.android.files.compat.ProxyFileDescriptorCallbackCompat$toProxyFileDescriptorCallback$1.onRead(ProxyFileDescriptorCallbackCompat.kt:8)
       at com.android.internal.os.FuseAppLoop.handleMessage(FuseAppLoop.java:181)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loopOnce(Looper.java:224)
       at android.os.Looper.loop(Looper.java:318)
       at android.os.HandlerThread.run(HandlerThread.java:67)
```

```
Fatal Exception: java.lang.NoSuchFieldError: No instance field $transform of type LE5/l; in class LI6/e; or its superclasses (declaration of 'I6.e' appears in /data/app/~~wmZ4dqUPFJD8T1vxbPFAYA==/me.zhanghai.android.files-xC8mCPOXXxDwi7kh2Bi6Og==/base.apk)
       at me.zhanghai.android.files.provider.sftp.client.FileByteChannel$onReadAsync$$inlined$map$1.get(FutureExtensions.kt:1)
       at me.zhanghai.android.files.provider.common.AbstractFileByteChannel$ReadBuffer.readIntoBuffer(AbstractFileByteChannel.kt:40)
       at me.zhanghai.android.files.provider.common.AbstractFileByteChannel$ReadBuffer.read(AbstractFileByteChannel.kt:40)
       at me.zhanghai.android.files.provider.common.AbstractFileByteChannel.read(AbstractFileByteChannel.kt:26)
       at me.zhanghai.android.files.provider.common.BaseDelegateSeekableByteChannel.read(DelegateSeekableByteChannel.kt:8)
       at me.zhanghai.android.files.file.FileProvider$ChannelCallback.onRead(FileProvider.kt:40)
       at me.zhanghai.android.files.compat.ProxyFileDescriptorCallbackCompat$toProxyFileDescriptorCallback$1.onRead(ProxyFileDescriptorCallbackCompat.kt:8)
       at com.android.internal.os.FuseAppLoop.handleMessage(FuseAppLoop.java:181)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loopOnce(Looper.java:205)
       at android.os.Looper.loop(Looper.java:294)
       at android.os.HandlerThread.run(HandlerThread.java:67)
```

**Comment by zhanghai** [comment]
https://issuetracker.google.com/issues/329250108

**Comment by pgaskin** [comment]
This is happening in [debug builds](https://github.com/pgaskin/MaterialFiles/actions/runs/8291635567) too:

```
03-15 01:25:59.565  3562  3613 E AndroidRuntime: FATAL EXCEPTION: FileProvider.CallbackThread
03-15 01:25:59.565  3562  3613 E AndroidRuntime: Process: me.zhanghai.android.files, PID: 3562
03-15 01:25:59.565  3562  3613 E AndroidRuntime: java.lang.NoSuchFieldError: No instance field $transform of type Lkotlin/jvm/functions/Function1; in class Lme/zhanghai/android/files/provider/sftp/client/FileByteChannel$onReadAsync$$inlined$map$1; or its superclasses (declaration of 'me.zhanghai.android.files.provider.sftp.client.FileByteChannel$onReadAsync$$inlined$map$1' appears in /data/app/~~n0iRUCQ004zyk5Fgqjo9Ow==/me.zhanghai.android.files-iATDfdYhLgJ3CmXBOXbhXg==/base.apk!classes7.dex)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.provider.sftp.client.FileByteChannel$onReadAsync$$inlined$map$1.get(FutureExtensions.kt:150)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.provider.common.AbstractFileByteChannel$ReadBuffer.readIntoBuffer(AbstractFileByteChannel.kt:230)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.provider.common.AbstractFileByteChannel$ReadBuffer.read(AbstractFileByteChannel.kt:211)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.provider.common.AbstractFileByteChannel.read(AbstractFileByteChannel.kt:49)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.provider.common.BaseDelegateSeekableByteChannel.read(DelegateSeekableByteChannel.kt:47)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.file.FileProvider$ChannelCallback.onRead(FileProvider.kt:287)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at me.zhanghai.android.files.compat.ProxyFileDescriptorCallbackCompat$toProxyFileDescriptorCallback$1.onRead(ProxyFileDescriptorCallbackCompat.kt:45)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at com.android.internal.os.FuseAppLoop.handleMessage(FuseAppLoop.java:181)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:205)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:294)
03-15 01:25:59.565  3562  3613 E AndroidRuntime: 	at android.os.HandlerThread.run(HandlerThread.java:67)
```

**Comment by pgaskin** [comment]
Manually inlining transformGet seems to be a sufficient [workaround](https://github.com/pgaskin/MaterialFiles/commit/d08b953da0cba492ac77ff95d4646190a7b90bbe) for now.

**Comment by zhanghai** [comment]
Yes, it seems a kotlinc issue instead of R8 issue...

**Comment by zhanghai** [comment]
This is actually https://youtrack.jetbrains.com/issue/KT-20245/NoSuchFieldError-on-usage-of-crossinline-lambda-in-anonymous-object-inside-another-anonymous-object-in-inline-function#focus=Comments-27-3104763.0-0 . I'll try to work around it in my code.

**Comment by Rorschach1010** [comment]
I tried the apk from the action artifacts but it didn't solve the problem for me. I'm not sure if the problem is the same as I don't understand the logs above.

When I want to start a video file, it doesn't load and sometimes even crashes. It works fine when the file is on my phone so the issue is just happening when I start something remotely via sftp or webdav. It used to work fine in the past.

To check if only Material Files has this issue, I tried the same with nova player, connected to sftp and tried the same files and nova player started them without issue.
Then I installed MiXplorer, connected to sftp and started the video files and they started without issues.

So this istn't an issue with the players I used but within MF only. As I said, it used to work in the past but I can't really tell you when the issue started to happen.

btw: Its not just media files. I also can't open (for example) pdfs on webdav and sftp. That's why I thought that this is the same issue but the artefact apk didn't solve it.

**Comment by zhanghai** [comment]
@Rorschach1010 Did you try the APK after my fix, e.g. the one in https://github.com/zhanghai/MaterialFiles/actions/runs/8292666999 ? I tried that it did allow me to open SFTP files.

And what was your WebDAV issue, any error message, stacktrace, or steps to reproduce?

**Comment by Rorschach1010** [comment]
> @Rorschach1010 Did you try the APK after my fix, e.g. the one in https://github.com/zhanghai/MaterialFiles/actions/runs/8292666999 ? I tried that it did allow me to open SFTP files.
> 
> And what was your WebDAV issue, any error message, stacktrace, or steps to reproduce?

Yes, I tried that apk and sftp files are still not working with the same error. The path to the file in the header (of my pdf viewer, audio player...) is shown like ip%3Aport%2F........ so it is converting the / slashes and the : colons. After a while MF shows the error: java.lang.ArrayIndexOutOfBoundsException: src.length=1 srcPos=0 dst.length=256 dstPos=.5 length=1


Webdav is opening pdfs and txt files more or less. Videos and music are opening but not really starting (it stays frozen) and after some time a toast appears with some java8.nio.file.FileSystemException: remote

Pictures aren't working at all (without error)

**Comment by Rorschach1010** [comment]
Seems like I'm able to get a stacktrace (which seems to be exactly the same as yours in the OP?) since it crashes sometimes when trying to start a video but this was with the 1.7.0 stable release. The new debug apk doesn't crash so I can't provide an error log

```
time: 1710621794247
msg: java.lang.NoSuchFieldError: No instance field $transform of type LE5/l; in class LI6/e; or its superclasses (declaration of 'I6.e' appears in /data/app/~~ffhFPoK_bAOCqVxjCpuiPg==/me.zhanghai.android.files-C92GJ7P9BY2_7-C-5lzidw==/base.apk)
stacktrace: java.lang.NoSuchFieldError: No instance field $transform of type LE5/l; in class LI6/e; or its superclasses (declaration of 'I6.e' appears in /data/app/~~ffhFPoK_bAOCqVxjCpuiPg==/me.zhanghai.android.files-C92GJ7P9BY2_7-C-5lzidw==/base.apk)
	at I6.e.get(SourceFile:1)
	at x6.a.read(SourceFile:40)
	at x6.e.read(SourceFile:26)
	at x6.o.read(SourceFile:8)
	at g6.a.b(SourceFile:40)
	at e6.n.onRead(SourceFile:8)
	at com.android.internal.os.FuseAppLoop.handleMessage(FuseAppLoop.java:181)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.os.HandlerThread.run(HandlerThread.java:67)
```

**Comment by zhanghai** [comment]
I see. Yes the crash is the bug this issue was about and that has been fixed. I could connect to https://test.rebex.net/ SFTP and view text files and images there without issue.

The path name is a red herring since the file name shown in your PDF viewer etc doesn't affect its content and is merely a reflection of the fact that the file is coming from another app via a URI.

The ArrayIndexOutOfBoundsException should be another sshj library issue that I wasn't able to fix.

WebDAV could be a server issue as I was able to playback video with good server implementations but was also stuck with a black screen for non-compliant servers. Which server are you using? And are there any more logs/stacktraces with the `java8.nio.file.FileSystemException: remote` message?

**Comment by Rorschach1010** [comment]
> I see. Yes the crash is the bug this issue was about and that has been fixed. I could connect to https://test.rebex.net/ SFTP and view text files and images there without issue.
> 
> The path name is a red herring since the file name shown in your PDF viewer etc doesn't affect its content and is merely a reflection of the fact that the file is coming from another app via a URI.
> 
> The ArrayIndexOutOfBoundsException should be another sshj library issue that I wasn't able to fix.
> 
> WebDAV could be a server issue as I was able to playback video with good server implementations but was also stuck with a black screen for non-compliant servers. Which server are you using? And are there any more logs/stacktraces with the `java8.nio.file.FileSystemException: remote` message?

Alright, I played with the debug build a little bit more. The good thing is, it really doesn't crash (so unfortunately I don't get a stacktrace). It seems like, when I wait long enough, it does open/start/play (depending on the file). However, it is so slow, it's not really usable tbh. that `java8.nio.file.FileSystemException:` (followed by the path) error is always showing as a toast, whenever I close the video/picture/pdf again. So something java related is slowing everything down. So it's most probably a dependency upstream issue, in this case as you said, sshj? I don't know anything about coding but since MiXplorer doesn't have this issue, there seems to be another way to implement ssh (without using sshj)?

I tried the test server you mentioned above. I didn't get any issues there but there are only txt and small png files but still, there is the possibility that it could be an issue with the sshd config of my. It doesn sound more like sshj is the issue, though.

**Comment by zhanghai** [comment]
sshj - yes, I'm thinking about switching to JSch in #1156.

**Comment by Rorschach1010** [comment]
> sshj - yes, I'm thinking about switching to JSch in #1156.

Nice, can't wait to test it so hopefully this solves this issue in the future.

In any way, your original issue from the OP is solved and already merged so this issue can be can be closed since the sshj issue isn't related to it.
Big thanks for your time and commitment.

**Comment by BenMcLean** [comment]
Will this be in the next release?

**Comment by zhanghai** [comment]
This has been fixed in 1.7.1

**Comment by BenMcLean** [comment]
Great! I'll be looking forward to when that shows up on [F-Droid](https://f-droid.org/en/packages/me.zhanghai.android.files/).
